File: leetcode/154-find-minimum-in-rotated-sorted-array-ii.rb
# @param {Integer[]} nums
# @return {Integer}
def find_min(nums)
  b = 0
  e = nums.size - 1
  min = nums[0]
  while b < e
    break if nums[b] < nums[e]
    mid = (e - b) / 2 + b
    if nums[mid] < nums[b]
      b += 1
      e = mid
    elsif nums[mid] > nums[b] or nums[b] != nums[e]
      b = mid + 1
    else
      while b < e and nums[b] == nums[e]
        b += 1
        e -= 1
      end
    end
  end
  [min, nums[b]].min
end

File: leetcode/154-find-minimum-in-rotated-sorted-array-ii.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int findMin(vector<int> &num) {
        return findMin(num, 0, num.size()-1);
    }
    int findMin(const vector<int> &num, int first, int last) {
        while (last > 0 && num[first] == num[last]) {
            --last;
        }
        if (first == last) {
            return num[first];
        }
        if (first == last - 1) {
            return min(num[first], num[last]);
        }
        if (num[first] < num[last]) {
            return num[first];
        }
        auto mid = first + (last - first + 1) /2;
        if (num[first] > num[mid]) {
            return findMin(num, first, mid);
        } else {
            return findMin(num, mid, last);
        }
    }
};
